hash.c

```diff
@@ -5913,10 +6007,26 @@ env_replace_i(VALUE key, VALUE val, VALUE keys)
 
 /*
  * call-seq:
- *   ENV.replace(hash) -> env
+ *   ENV.replace(hash) -> ENV
+ *
+ * Replaces the entire content of the environment variables
+ * with the name/value pairs in the given +hash+;
+ * returns ENV.
  *
- * Replaces the contents of the environment variables with the contents of
- * +hash+.
+ * Replaces the content of ENV with the given pairs:
+ *   ENV.replace('foo' => '0', 'bar' => '1') # => ENV
+ *   ENV.to_hash # => {"bar"=>"1", "foo"=>"0"}
+ *
+ * Raises an exception if a name or value is invalid.
+ * See {Invalid Names and Values}[#class-ENV-label-Invalid-Names+and+Values].
+ *
+ * Adds new environment variables up to the first error encountered:
```

Comment:
I'm not sure we can/should define the result after an error.

Reply:
For other errors, we say nothing about the state of ENV, implying that it's not changed (which is true).  Here, the state of ENV -is- changed, and I think we have to say so.

Actually, I think this is a bug.  On an error, ENV should never be changed.

Reply:
Demo:

irb
ENV.clear
# => {}
ENV.replace('foo' => '0', Object.new => Object.new)
Traceback (most recent call last):
        5: from C:/Ruby26-x64/bin/irb.cmd:31:in `<main>'
        4: from C:/Ruby26-x64/bin/irb.cmd:31:in `load'
        3: from C:/Ruby26-x64/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        2: from (irb):2
        1: from (irb):2:in `replace'
TypeError (no implicit conversion of Object into String)
ENV
# => {"foo"=>"0"}

Reply:
Also `ENV.update` has same behavior, and I think it is hard to guarantee your expectation as the change would be expected immediately in its block form.

Reply:
So, ready to merge now?